### DISCUSSION TITLE
Crossbar dependency issues with python 3.12 // crossbar project

### DISCUSSION THREAD

**Discussion by mangelajo** [labgrid-project/labgrid#1335] [General]
Hey, I am Miguel Ángel, from https://jumpstarter.dev a really small project compared to yours; I presented it at last FOSDEM and people in the room directed me to look into LabGrid, which I am evaluating, it seems much more mature, with many more capabilities compared to what we built in a few months.

I was trying to setup the coordinator to test the distributed environment, and I hit an issue with crossbar:

  File "/home/ajo/labgrid/crossbar-venv/lib/python3.12/site-packages/crossbar/common/reloader.py", line 40, in <module>
    from imp import reload
ModuleNotFoundError: No module named 'imp'

imp was deprecated, and I could switch it to importlib

Then I hit: 

  File "/home/ajo/labgrid/crossbar-venv/lib/python3.12/site-packages/urllib3/exceptions.py", line 2, in <module>
    from .packages.six.moves.http_client import (
ModuleNotFoundError: No module named 'urllib3.packages.six.moves'

which I still need to figure out.

Then I looked at the crossbar project, which seemed inactive for the last 7 months, and their main website is down.
I guess crossbar is deeply ingrained in the project, is it possible to replace it? are there any action plans?

Best regards, and congratulations for the project.

**Comment by Emantor** [comment]
Yes, `crossbar` as a dependency is becoming a real issue. I have started to experiment with grpc, which has the upside that our startup overhead for client calls is also lower. Unfortunately this requires us to implement our own session handling. Long term we want to move away from crossbar.

**Reply by flxzt** [reply]
+1 for gRPC. It would also open up the ability to hook into labgrid from e.g. a UI, monitoring service, .. more easily with it's many different language bindings.

**Comment by Bastian-Krause** [comment]
For the state of crossbar Python 3.12 compatibility, see #1260.

**Reply by mangelajo** [reply]
I have tested #1260 and it worked, this will help me continue the evaluation. 👍
